elastic/detection-rules
# [FR] Add CLI Command for Prebuilt Rules Release Package Update Details

## Related
* https://github.com/elastic/detection-rules/pull/2029

## Overview

This pull request adds additional functionality for exploring differences between prebuilt rule release packages. _Keep in mind, this functionality is only intended for internal use when doing release updates, but is accessible to everyone._ Below is a list of notes for this:

## `package-updates`
- Rules are pulled from EPR as specified by `--versions`. This is meant to be comparative so only two packages are allowed to be compared.
- Since rules are pulled from EPR, they are post-validation and thus do not need to be loaded back through the rule loader or hardcoded rule dataclasses
- `details` option allows a use to specify which diff details they are seeking, such as only new rules, MITRE ATT&CK coverage updates, query adjustments, data sources, etc. While there is more we could add, these are the basics o understanding what updates to communicate when releasing prebuilt rules.
- `tabulate` was added as a dependency for easy table output to the console; this was done in favor of pandas as the metrics are not difficult to determine, nor do we need such as large dependency

### Testing
1. Clone the repository
2. Checkout branch
3. Run `python -m detection_rules dev diff versions 8.12.2 8.12.3 --summary`
4. Run `python -m detection_rules dev diff versions 8.12.2 8.12.3 --details [OPTION]` (run all the optional parameters)

<details><summary>Expected Output</summary>
<img width="708" alt="Screenshot 2024-01-29 at 11 36 52 PM" src="https://github.com/elastic/detection-rules/assets/99630311/834714fd-e9a5-4371-a085-df9d09aad7d3">
<img width="675" alt="Screenshot 2024-01-29 at 11 36 09 PM" src="https://github.com/elastic/detection-rules/assets/99630311/0dbb940e-ae60-4589-a05d-24e37e8696fd">
<img width="1140" alt="Screenshot 2024-01-29 at 11 35 29 PM" src="https://github.com/elastic/detection-rules/assets/99630311/e3b82522-e07b-4fd4-98f0-4174f514ca05">
<img width="628" alt="Screenshot 2024-01-29 at 11 34 41 PM" src="https://github.com/elastic/detection-rules/assets/99630311/23516880-26b3-4b3f-8d98-e1a3e43e7752">
<img width="598" alt="Screenshot 2024-01-29 at 7 05 08 PM" src="https://github.com/elastic/detection-rules/assets/99630311/d25712ae-7df8-4b7a-afda-272e235618cc">
<img width="669" alt="Screenshot 2024-01-29 at 6 45 19 PM" src="https://github.com/elastic/detection-rules/assets/99630311/31478e6d-27d6-44a1-80c1-8fec07e95f75">
</details>

## `check-available-updates`
- Loads all prebuilt detection rule packages from EPR
- Iterates manifests to abstract package version and compatible Kibana version
- Formats to tabulate table and shows output

This command is useful for releases when attempting to identify previous packages manually, rather than sending custom CURL requests to the EPR API.

#### Testing
1. Clone the repository
2. Checkout branch
3. Run `python -m detection_rules dev check-available-packages`

<details><summary>Expected Output</summary>
<img width="925" alt="Screenshot 2024-01-30 at 10 31 33 AM" src="https://github.com/elastic/detection-rules/assets/99630311/898d4655-320b-4fcc-91c0-e7be025cbcc0">
</details>
